### ISSUE TITLE
The loss and dice_coef are not consistent to each other. 

### ISSUE THREAD

**Issue by salan668** [ellisdg/3DUnetCNN#33]
Hi, I tried to use this UNet to segment the image and used Dice as the loss function. The function was like below and I didn't change it. 

```
def dice_coef(y_true, y_pred, smooth=1.):
    y_true_f = K.flatten(y_true)
    y_pred_f = K.flatten(y_pred)
    intersection = K.sum(y_true_f * y_pred_f)
    return (2. * intersection + smooth) / (K.sum(y_true_f) + K.sum(y_pred_f) + smooth)


def dice_coef_loss(y_true, y_pred):
    return -dice_coef(y_true, y_pred)
```
The complie is like:
```
model.compile(optimizer=Adam(lr=initial_learning_rate), loss=dice_coef_loss, metrics=[dice_coef])
```

However, during the training, the log message is like below:
```
Epoch 1/1000
299s - loss: 6.4668 - dice_coef: 0.2710 - val_loss: 5.3623 - val_dice_coef: 0.5395
Epoch 2/1000
228s - loss: 4.7529 - dice_coef: 0.5754 - val_loss: 4.4627 - val_dice_coef: 0.3444
Epoch 3/1000
229s - loss: 3.7705 - dice_coef: 0.6485 - val_loss: 3.3135 - val_dice_coef: 0.7470
Epoch 4/1000
232s - loss: 3.0341 - dice_coef: 0.7334 - val_loss: 2.7504 - val_dice_coef: 0.7445
Epoch 5/1000
230s - loss: 2.5172 - dice_coef: 0.7546 - val_loss: 2.2995 - val_dice_coef: 0.7674
Epoch 6/1000
230s - loss: 2.0970 - dice_coef: 0.7988 - val_loss: 1.9096 - val_dice_coef: 0.8277
```

In my opinion, I think the loss should be negative dice_coef. If I'm wrong, how was this loss calculated? 
Further, Somethings the loss is equal to negativedice_coef, sometimes it is not. 

Thank you.

**Comment by ellisdg** [comment]
When you start training are you sure it is not using a saved model instead? That is the only explanation I can think of. As you said, the loss should just be the negative of the dice coefficient. I don't see how that could change. In my BRATS training script, if a model already exists, the script will [load that model instead of creating a new one](https://github.com/ellisdg/3DUnetCNN/blob/master/brats/train.py#L32). Is it possible that is happening with your model?

**Comment by salan668** [comment]
Thanks for your kind reply. 

I didn't load the saved model. I just trained a new one. I changed the code based on yours and there is on script to load the saved model. what I do in the script is just like below:

```
from model import unet_model_3d
model = unet_model_3d(input_shape, downsize_filters_factor=1, deconvolution=True)
result = model.fit_generator(train_generator, steps_per_epoch=100, epochs=1000, verbose=2, validation_data=test_generator, validation_steps=20, callbacks=cbks)
```

Another curious apperance is that the loss always drops and never increases. There is no minor concussion. I can't explain that.

**Comment by salan668** [comment]
Hi, I know the reason. I added the regulization in the training. So the weight term was added in the dice coef. Thank you. Let me close it

**Comment by ellisdg** [comment]
Thanks for the update! I was really confused as to how that could happen, but that makes sense now.
